source/common/common/inline_map.h

```diff
@@ -0,0 +1,432 @@
+ * filter state always uses the filter name as the key and the filter name is known at compile time.
+ * By using the handle, the filter state could get the key/value pair without any hash.
+ *
+ * This class also provides the interface to access the key/value pair by the normal key and the
+ * interface has similar searching performance as the normal hash map. But the insertion and removal
+ * by the normal key is slower than the normal hash map.
+ *
+ *
+ * These is usage example:
+ *
+ *   // Define descriptor to store the inline keys. The descriptor should have lifetime that no
+ *   // shorter than all inline maps created by the descriptor. You can create static and global
+ *   // descriptor by ``MUTABLE_CONSTRUCT_ON_FIRST_USE`` for cross-modules usage, but it is not
+ *   // required. A local descriptor is also fine if the inline map is only used in the local scope.
+ *   InlineMapDescriptor<std::string> descriptor;
+ *
+ *   // Create the handle for inline key. We should never do this after bootstrapping. Typically, we
+ *   // can do this by define a global variable.
+ *   InlineMapDescriptor<std::string>::Handle handle = descriptor.addHandle("inline_key");
+ *
+ *   // Finalize the descriptor. No further changes are allowed to the descriptor after this point.
+ *   descriptor.finalize();
+ *
+ *   // Create the inline map.
+ *   InlineMapPtr<std::string, std::string> inline_map = InlineMap<std::string,
+ *     std::string>::create(descriptor);
+ *
+ *   // Get value by handle.
+ *   inline_map->insert(handle, "value");
+ *   EXPECT_EQ(*inline_map->lookup(handle), "value");
+ */
+
+/**
+ * Inline map descriptor to store the inline keys and mapping of the inline keys to the inline
+ * id.
+ */
+template <class Key> class InlineMapDescriptor {
+public:
+  /**
+   * This is the handle used to access the key/value pair in the inline map. The handle is
+   * lightweight and could be copied and passed by value.
+   */
+  class Handle {
+  public:
+    /**
+     * Get the id of the inline entry in the inline array. This could be used to access the
+     * key/value pair in the inline map without hash searching.
+     */
+    uint64_t inlineId() const { return inline_id_; }
+
+  private:
+    friend class InlineMapDescriptor;
+
+    // This constructor should only be called by InlineMapDescriptor.
+    Handle(uint64_t inline_id) : inline_id_(inline_id) {}
+
+    uint64_t inline_id_{};
+  };
+
+  using InlineKeysMap = absl::flat_hash_map<Key, Handle>;
+  using InlineKeys = std::vector<Key>;
+
+  InlineMapDescriptor() = default;
+
+  /**
+   * Add an inline key and return related handle. May only be called before finalize().
+   */
+  template <class ArgK> Handle addInlineKey(const ArgK& key) {
+    RELEASE_ASSERT(!finalized_, "Cannot create new inline key after finalize()");
+
+    if (auto it = inline_keys_map_.find(key); it != inline_keys_map_.end()) {
+      // If the key is already added, return related inline handle directly.
```

Comment:
These inline keys may be configable (depend on the usage of the descriptor and inline map). And maybe some inline keys are come from configuration, some other inline keys are come from hard-code. It's very possible that users configured a key that has added by the hard-code.

So considering this case, rather than throw an error, I prefer to just return the same index if repeat keys are added.

Reply:
ok